src/main/java/org/jboss/threads/EnhancedQueueExecutor.java

```diff
@@ -1727,79 +1716,70 @@ boolean doStartThread(Runnable runnable) throws RejectedExecutionException {
+                    // additional success postconditions: -
+                    // failure postconditions: -
                     // post-actions (succeed):
-                    //   unpark thread and return
+                    //   run state change ex2
                     // post-actions (fail):
-                    //   retry outer with new tail(snapshot)
-                    if (consumerNode.compareAndSetTask(WAITING, runnable)) {
-                        consumerNode.unpark();
-                        result = EXE_OK;
-                        break;
+                    //   retry with new tail(snapshot)
+                    if (tail.compareAndSetNext(tailNext, tailNextNext)) {
+                        PoolThreadNode consumerNode = (PoolThreadNode) tailNext;
+                        // state change ex2:
+                        //   tail(snapshot).next(snapshot).task ← runnable
+                        // succeeds: ex1
+                        // preconditions:
+                        //   tail(snapshot).next(snapshot).task = WAITING
+                        // post-actions (succeed):
+                        //   unpark thread and return
+                        // post-actions (fail):
+                        //   retry outer with new tail(snapshot)
+                        if (consumerNode.compareAndSetTask(WAITING, runnable)) {
+                            // unpark must be executed without a lock
+                            toUnpark = consumerNode;
+                            result = EXE_OK;
+                            break;
+                        }
+                        // otherwise the consumer gave up or was exited already, so fall out and...
                     }
-                    // otherwise the consumer gave up or was exited already, so fall out and...
-                }
-                // retry with new tail(snapshot) as was foretold
-                tail = this.tail;
-                if (UPDATE_STATISTICS) spinMisses.increment();
-            } else if (tailNext == null) {
-                // no consumers available; maybe we can start one
-                int tr = tryAllocateThread(growthResistance);
-                if (tr == AT_YES) {
-                    result = EXE_CREATE_THREAD;
-                    break;
-                }
-                if (tr == AT_SHUTDOWN) {
-                    result = EXE_REJECT_SHUTDOWN;
-                    break;
-                }
-                assert tr == AT_NO;
-                // no; try to enqueue
-                if (! NO_QUEUE_LIMIT && ! increaseQueueSize()) {
-                    // queue is full
-                    // OK last effort to create a thread, disregarding growth limit
-                    tr = tryAllocateThread(0.0f);
+                    // retry with new tail(snapshot) as was foretold
+                    tail = this.tail;
+                    if (UPDATE_STATISTICS) spinMisses.increment();
```

Comment:
hmm, while we're here it probably makes sense to increment stats prior to updating tail for a tight spin.